NSrtm.Core/Pgm/PgmDataDescription.cs

```diff
@@ -0,0 +1,186 @@
+                hashCode = (hashCode * 397) ^ _gridGraphHeightPoints;
+                hashCode = (hashCode * 397) ^ _maxValue;
+                hashCode = (hashCode * 397) ^ _preambleLength;
+                return hashCode;
+            }
+        }
+
+        public double Offset { get { return _offset; } }
+        public double Scale { get { return _scale; } }
+        public int OriginLat { get { return _originLat; } }
+        public int OriginLon { get { return _originLon; } }
+        public int GridGraphWidthPoints { get { return _gridGraphWidthPoints; } }
+        public int GridGraphHeightPoints { get { return _gridGraphHeightPoints; } }
+        public int MaxValue { get { return _maxValue; } }
+        public int NumberOfPoints { get { return _numberOfPoints; } }
+        public int PreambleLength { get { return _preambleLength; } }
+        public double LatitudeIncrement { get { return _latitudeIncrement; } }
+        public double LongitudeIncrement { get { return _longitudeIncrement; } }
+
+        public bool Equals(PgmDataDescription other)
+        {
+            return _offset.Equals(other._offset) && _scale.Equals(other._scale) && _originLat == other._originLat && _originLon == other._originLon &&
+                   _gridGraphWidthPoints == other._gridGraphWidthPoints && _gridGraphHeightPoints == other._gridGraphHeightPoints &&
+                   _maxValue == other._maxValue && _preambleLength == other._preambleLength;
+        }
+
+        #region Static Members
+
+        public static bool operator ==(PgmDataDescription left, PgmDataDescription right)
+        {
+            return left.Equals(right);
+        }
+
+        public static bool operator !=(PgmDataDescription left, PgmDataDescription right)
+        {
+            return !left.Equals(right);
+        }
+
+        #endregion
+    }
+
+    internal static class PgmRawDataConverter
+    {
+        public static double ToEgmFormat(this UInt16 rawData, PgmDataDescription pgmParameters)
+        {
+            if (rawData > pgmParameters.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException("rawData");
+            }
+            return rawData * pgmParameters.Scale + pgmParameters.Offset;
+        }
+    }
+
+    internal static class PgmDataDescriptionExtractor
+    {
+        private static readonly Regex LatLonParser = new Regex("^(?<deg>[-+0-9]+)+(?<pos>[ENSW]*)$");
+
+        public static PgmDataDescription FromStream(Stream stream)
+        {
+            var preamble = extractPreambleFromStream(stream);
+            return getConstatantsFromPreamble(preamble);
+        }
+
+        private static string extractPreambleFromStream(Stream stream)
+        {
+            using (var reader = new StreamReader(stream, Encoding.UTF8, false, 1024, true))
+            {
+                const string _pgmMagicNumber = "P5";
+                var preamble = reader.ReadLine();
```

Comment:
If you are sure this pattern is bettern than processing file line by line at least use `StringBuilder` here.